File: sdk/go/worker_trust_verify.go
package capsdk

import (
	"crypto/ecdsa"
	"fmt"
	"time"

	agentv1 "github.com/cordum-io/cap/v2/cordum/agent/v1"
	"google.golang.org/protobuf/proto"
)

// VerifyWorkerHandshakeChallenge authenticates and correlates a scheduler reply.
func VerifyWorkerHandshakeChallenge(config *WorkerTrustConfig, request, response *agentv1.BusPacket, now time.Time) (*VerifiedWorkerHandshakeChallenge, error) {
	if err := ValidateWorkerTrustConfig(config); err != nil {
		return nil, err
	}
	if err := validateChallengePackets(config, request, response, now); err != nil {
		return nil, err
	}
	return &VerifiedWorkerHandshakeChallenge{
		request: proto.Clone(request).(*agentv1.BusPacket), response: proto.Clone(response).(*agentv1.BusPacket),
	}, nil
}

// VerifyWorkerHandshakeResult authenticates a correlated signed result.
func VerifyWorkerHandshakeResult(config *WorkerTrustConfig, verified *VerifiedWorkerHandshakeChallenge, authenticate, response *agentv1.BusPacket, now time.Time) (*WorkerHandshakeSession, error) {
	if err := ValidateWorkerTrustConfig(config); err != nil {
		return nil, err
	}
	if err := validateResultPackets(config, verified, authenticate, response, now); err != nil {
		return nil, err
	}
	result := response.GetWorkerHandshakeResult()
	if !result.GetAccepted() {
		return nil, &WorkerHandshakeRejectionError{Reason: result.GetRejectionReason()}
	}
	issuedAt, _ := checkedProtoTime(result.GetIssuedAt())
	expiresAt, _ := checkedProtoTime(result.GetTokenExpiresAt())
	return &WorkerHandshakeSession{Token: response.GetAuthToken(), IssuedAt: issuedAt, ExpiresAt: expiresAt}, nil
}

func validateChallengePackets(config *WorkerTrustConfig, request, response *agentv1.BusPacket, now time.Time) error {
	if err := ValidateWorkerTrustPacket(request); err != nil {
		return fmt.Errorf("%w: request: %w", ErrWorkerHandshakePacket, err)
	}
	workerKeys := map[string]*ecdsa.PublicKey{config.ProofKeyID: &config.ProofPrivateKey.PublicKey}
	if err := VerifyTrustHandshake(request, workerKeys); err != nil {
		return fmt.Errorf("%w: request signature: %w", ErrWorkerHandshakePacket, err)
	}
	if err := validateSignedSchedulerPacket(config, response, now); err != nil {
		return err
	}
	challenge := response.GetWorkerHandshakeChallenge()
	if err := correlateChallengeRequest(config, request.GetWorkerHandshakeChallengeRequest(), challenge); err != nil {
		return err
	}
	if err := validateChallengeFreshness(challenge, now); err != nil {
		return err
	}
	return nil
}

func validateResultPackets(config *WorkerTrustConfig, verified *VerifiedWorkerHandshakeChallenge, authenticate, response *agentv1.BusPacket, now time.Time) error {
	if verified == nil || verified.request == nil || verified.response == nil {
		return fmt.Errorf("%w: challenge is unverified", ErrWorkerHandshakeBinding)
	}
	if err := ValidateWorkerTrustPacket(authenticate); err != nil {
		return fmt.Errorf("%w: authenticate: %w", ErrWorkerHandshakePacket, err)
	}
	workerKeys := map[string]*ecdsa.PublicKey{config.ProofKeyID: &config.ProofPrivateKey.PublicKey}
	if err := VerifyTrustHandshake(authenticate, workerKeys); err != nil {
		return fmt.Errorf("%w: authenticate signature: %w", ErrWorkerHandshakePacket, err)
	}
	if err := validateSignedSchedulerPacket(config, response, now); err != nil {
		return err
	}
	want := verified.response.GetWorkerHandshakeChallenge()
	got := response.GetWorkerHandshakeResult().GetChallenge()
	if !proto.Equal(want, got) || !proto.Equal(want, authenticate.GetWorkerHandshakeAuthenticate().GetChallenge()) {
		return fmt.Errorf("%w: result challenge differs", ErrWorkerHandshakeBinding)
	}
	if err := validateChallengeFreshness(got, now); err != nil {
		return err
	}
	result := response.GetWorkerHandshakeResult()
	if err := validateTimestampSkew(result.GetIssuedAt(), now, "result.issued_at"); err != nil {
		return err
	}
	if result.GetAccepted() {
		expiresAt, _ := checkedProtoTime(result.GetTokenExpiresAt())
		if !expiresAt.After(now.UTC()) {
			return fmt.Errorf("%w: result token is already expired", ErrWorkerHandshakeExpired)
		}
		if got.GetPurpose() == renewWorkerHandshakePurpose() && response.GetAuthToken() == authenticate.GetAuthToken() {
			return fmt.Errorf("%w: renewal did not rotate session token", ErrWorkerHandshakeBinding)
		}
	}
	return nil
}

func validateSignedSchedulerPacket(config *WorkerTrustConfig, packet *agentv1.BusPacket, now time.Time) error {
	if err := ValidateWorkerTrustPacket(packet); err != nil {
		return fmt.Errorf("%w: scheduler packet: %w", ErrWorkerHandshakePacket, err)
	}
	if packet.GetSenderId() != config.ExpectedSchedulerID {
		return fmt.Errorf("%w: scheduler sender differs", ErrWorkerHandshakeBinding)
	}
	if err := validateTimestampSkew(packet.GetCreatedAt(), now, "created_at"); err != nil {
		return err
	}
	if err := VerifyTrustHandshake(packet, config.SchedulerPublicKeys); err != nil {
		return fmt.Errorf("%w: scheduler signature: %w", ErrWorkerHandshakePacket, err)
	}
	return nil
}

func correlateChallengeRequest(config *WorkerTrustConfig, request *agentv1.WorkerHandshakeChallengeRequest, challenge *agentv1.WorkerHandshakeChallenge) error {
	if request == nil || challenge == nil {
		return fmt.Errorf("%w: request or challenge missing", ErrWorkerHandshakeBinding)
	}
	matches := challenge.GetRequestId() == request.GetRequestId() && challenge.GetTraceId() == request.GetTraceId() &&
		challenge.GetWorkerId() == request.GetWorkerId() && challenge.GetProofKeyId() == request.GetProofKeyId() &&
		challenge.GetProofAlgorithm() == request.GetProofAlgorithm() && challenge.GetAudience() == request.GetAudience() &&
		challenge.GetPurpose() == request.GetPurpose() && sameNonce(challenge.GetClientNonce(), request.GetClientNonce()) &&
		challenge.GetProtocolVersion() == request.GetProtocolVersion() && challenge.GetSdkVersion() == request.GetSdkVersion()
	if !matches || challenge.GetAgentId() != config.ExpectedAgentID || challenge.GetTenantId() != config.TenantID {
		return fmt.Errorf("%w: challenge does not match request or expected identity", ErrWorkerHandshakeBinding)
	}
	return nil
}

func validateChallengeFreshness(challenge *agentv1.WorkerHandshakeChallenge, now time.Time) error {
	if err := validateTimestampSkew(challenge.GetIssuedAt(), now, "challenge.issued_at"); err != nil {
		return err
	}
	expiresAt, ok := checkedProtoTime(challenge.GetExpiresAt())
	if !ok || !now.UTC().Before(expiresAt) {
		return fmt.Errorf("%w: challenge expired", ErrWorkerHandshakeExpired)
	}
	return nil
}

File: sdk/python/cap/worker_trust_verify.py
"""Verification and correlation for worker trust-handshake responses."""

from cap.trust_signing import TrustSigningError, verify_trust_packet
from cap.worker_trust import (
    WORKER_HANDSHAKE_MAX_SKEW,
    VerifiedWorkerHandshakeChallenge,
    WorkerHandshakeBindingError,
    WorkerHandshakeExpiredError,
    WorkerHandshakePacketError,
    WorkerHandshakeRejectionError,
    WorkerHandshakeSession,
    _renew_purpose,
    _serialize,
    _timestamp_datetime,
    _utc,
    _verified_packets,
    validate_worker_trust_config,
)
from cap.worker_trust_validate import validate_worker_trust_packet


def verify_challenge(config, request, response, now):
    """Authenticate and correlate one scheduler challenge response."""

    validate_worker_trust_config(config)
    validate_worker_trust_packet(request)
    try:
        verify_trust_packet(
            request,
            {config.proof_key_id: config.proof_private_key.public_key()},
        )
    except TrustSigningError as exc:
        raise WorkerHandshakePacketError("request signature is invalid") from exc
    _validate_skew(request.created_at, now, "request.created_at")
    _verify_scheduler_packet(config, response, now)
    _correlate_challenge(config, request, response)
    _validate_challenge_freshness(response.worker_handshake_challenge, now)
    return VerifiedWorkerHandshakeChallenge._create(request, response)


def verify_result(config, verified, authenticate, response, now):
    """Authenticate a correlated result and return only a live session."""

    validate_worker_trust_config(config)
    _verified_packets(verified)
    validate_worker_trust_packet(authenticate)
    try:
        verify_trust_packet(
            authenticate,
            {config.proof_key_id: config.proof_private_key.public_key()},
        )
    except TrustSigningError as exc:
        raise WorkerHandshakePacketError("authenticate signature is invalid") from exc
    _verify_scheduler_packet(config, response, now)
    _correlate_result(verified, authenticate, response)
    result = response.worker_handshake_result
    _validate_challenge_freshness(result.challenge, now)
    _validate_skew(result.issued_at, now, "result.issued_at")
    if not result.accepted:
        raise WorkerHandshakeRejectionError(result.rejection_reason)
    expires_at = _timestamp_datetime(result.token_expires_at, "token_expires_at")
    if expires_at <= _utc(now):
        raise WorkerHandshakeExpiredError("result token is already expired")
    if result.challenge.purpose == _renew_purpose() and response.auth_token == authenticate.auth_token:
        raise WorkerHandshakeBindingError("renewal did not rotate session token")
    issued_at = _timestamp_datetime(result.issued_at, "issued_at")
    return WorkerHandshakeSession(response.auth_token, issued_at, expires_at)


def _verify_scheduler_packet(config, packet, now):
    validate_worker_trust_packet(packet)
    if packet.sender_id != config.expected_scheduler_id:
        raise WorkerHandshakeBindingError("scheduler sender differs")
    _validate_skew(packet.created_at, now, "created_at")
    try:
        verify_trust_packet(packet, config.scheduler_public_keys)
    except TrustSigningError as exc:
        raise WorkerHandshakePacketError("scheduler signature is invalid") from exc


def _correlate_challenge(config, request_packet, response_packet):
    request = request_packet.worker_handshake_challenge_request
    challenge = response_packet.worker_handshake_challenge
    fields = (
        "request_id", "trace_id", "worker_id", "proof_key_id",
        "proof_algorithm", "audience", "purpose", "client_nonce",
        "protocol_version", "sdk_version",
    )
    if any(getattr(challenge, field) != getattr(request, field) for field in fields):
        raise WorkerHandshakeBindingError("challenge does not match request")
    if (challenge.agent_id != config.expected_agent_id or
            challenge.tenant_id != config.tenant_id):
        raise WorkerHandshakeBindingError("challenge identity differs")


def _correlate_result(verified, authenticate, response):
    wanted = _serialize(_verified_packets(verified)[1].worker_handshake_challenge)
    got = _serialize(response.worker_handshake_result.challenge)
    sent = _serialize(authenticate.worker_handshake_authenticate.challenge)
    if got != wanted or sent != wanted:
        raise WorkerHandshakeBindingError("result challenge differs")


def _validate_challenge_freshness(challenge, now):
    _validate_skew(challenge.issued_at, now, "challenge.issued_at")
    expires_at = _timestamp_datetime(challenge.expires_at, "challenge.expires_at")
    if _utc(now) >= expires_at:
        raise WorkerHandshakeExpiredError("challenge expired")


def _validate_skew(timestamp, now, field):
    delta = _timestamp_datetime(timestamp, field) - _utc(now)
    if delta > WORKER_HANDSHAKE_MAX_SKEW or delta < -WORKER_HANDSHAKE_MAX_SKEW:
        raise WorkerHandshakeExpiredError("{} outside allowed skew".format(field))


__all__ = ["verify_challenge", "verify_result"]
